### ISSUE TITLE
Identical non-String values cannot be independently changed

### ISSUE THREAD

**Issue by raz3k** [wwkimball/yamlpath#44]
## Operating System

1. Name/Distribution: CentOS
2. Version: pre-7.7

## Version of Python and packages in use at the time of the issue.

1. [Distribution](https://wiki.python.org/moin/PythonDistributions): Python 3.6 from EPEL
2. Python Version: Python 3.6.8
3. Version of yamlpath installed: 2.3.0
4. Version of ruamel.yaml installed: 0.16.5

## Minimum sample of YAML (or compatible) data necessary to trigger the issue
```
---
somestring:
  string:
    someotherstring: true

otherstring:
  default:
    config:
      deploy:
        me: true
```
## Complete steps to reproduce the issue when triggered via:
```
from yamlpath.exceptions import YAMLPathException
from yamlpath.func import get_yaml_data, get_yaml_editor
from yamlpath import Processor
from yamlpath.wrappers import ConsolePrinter
from ruamel.yaml import YAML

class Args:
  debug = False
  verbose = False
  quiet = True

instance = YAML()

instance.default_flow_style = False
instance.explicit_start = True

yaml = get_yaml_editor()

log = ConsolePrinter(Args())
yaml_data = get_yaml_data(yaml, log, "/tmp/input.yaml")

processor = Processor(log, yaml_data)
try:
    processor.set_value("otherstring.default.config.deploy.me", "set_value")
except YAMLPathException as ex:
    log.critical(ex, 119)

with open("/tmp/output.yaml", 'w') as yaml_dump:
       instance.dump(yaml_data, yaml_dump)
```
## Expected Outcome
```
---
somestring:
  string:
    someotherstring: true

otherstring:
  default:
    config:
      deploy:
        me: set_value
```
## Actual Outcome
```
---
somestring:
  string:
    someotherstring: set_value

otherstring:
  default:
    config:
      deploy:
        me: set_value
```

**Comment by wwkimball** [comment]
That's certainly unexpected!  I actually use this project quite a lot and I never see this behavior.  There's something special about this specific case and I'd wager it has something to do with the transformation of a Boolean (true) to a String (set_value) value.  I suspect the underlying library or Python itself are marking those Boolean values as being two instances of one value, so when you change one, they both change (because they are just two pointers to the same value).

I'll get to the bottom of this and get a fix out ASAP.  Thanks for finding *and reporting* this!!

**Comment by wwkimball** [comment]
I've changed the issue title because this issue isn't what it seems.

I have found that the issue is *not* a problem with YAMLPath itself.  In fact, you seem to be mixing up data-types and that's causing your use-case to trip into one of ruamel.yaml or Python's aggressive efficiency traits.  Left to their own devices, ruamel.yaml or Python collapses all like values as references to each other in an aggressive pursuit of memory conservation.  While you and I might think of Boolean values as being a mere bit of data not worthy of pointer overhead, ruamel.yaml or Python appears to be considering them worth treating as references to one another when their values are identical.  So, every Boolean `true` within the YAML data are treated as references to each other and every Boolean `false` are similarly treated.  This is obviously not desirable when we intend to independently change one of these Boolean states.

When dealing with actual String values, everything works as expected.  However, you are providing Boolean values where you are expecting Strings (based on your key names).  You see, `true` is Boolean whereas `"true"` and `'true'` are Strings (note the `"` and `'` demarcation symbols).  If I wrap even one of your values in String demarcation symbols, there is no unexpected change to the other key.  So, use actual String values when you mean to use Strings, and there is no issue.

That said, anyone who intends to use real Booleans will face a nasty issue here.  So, this still needs an urgent fix.  I don't yet know whether this is being caused by ruamel.yaml or Python.  I'll keep digging and see whether this needs an Issue report against ruamel.yaml, Python itself, or whether I can somehow circumvent this overly aggressive optimization within YAMLPath.

**Comment by wwkimball** [comment]
After some more testing, I have updated the title of this Issue again and retracted my former comment.  I have discovered that identical Boolean and Integer values somehow become entangled and are updated together whenever users attempt to change one or another. Identical String values do not suffer this entanglement.  This is really bad and I'll get a fix out ASAP.  For as much as I use this project myself, I am amazed that I haven't encountered this before today!

**Comment by raz3k** [comment]
Thank you so much for taking the time to investigate this issue. Hope you will able to get a fix soon.

**Comment by wwkimball** [comment]
I discovered the issue is with some code that was contributed to my project by the author of ruamel.yaml, a library which YAMLPath relies heavily upon.  At his request when I asked about this matter, I have opened a [StackOverflow Question](https://stackoverflow.com/questions/58118589/ruamel-yaml-recurse-function-unexpectedly-updates-every-instance-of-non-string) to explain what is happening and request further help from him.  When we have a solution, I'll update YAMLPath accordingly.

**Comment by raz3k** [comment]
@wwkimball FYI: The author answered.

**Comment by wwkimball** [comment]
Yes and the suggested answer will require a substantial refactoring of my code.  I don't presently track each node's parent as I traverse the object tree.  I will need some time to implement this fix, debug, update unit tests, and publish a new release.  Please expect a 3-4 week delay.

If you or anyone else can afford more time to this than I can, please feel free to open a PR with this fix!  :)

**Comment by wwkimball** [comment]
Just an update:  work on this issue is slowly progressing.  You can track the work-in-progress via the bugfix/44 branch.

**Comment by adabuleanu** [comment]
Any updates on this? I was able to hit it in production as well.

**Comment by wwkimball** [comment]
Just and update:  I took a _long_ break from all Open Source projects for the holidays.  I'm back in and have been working on this issue.  I have incorporated Anthon's bug fix, which has solved this issue.  However, because there was so much refactoring required to feed new parameters for the new code, I've had to rework a ton of my code.

At this time, everything is working **except** Collectors.  If you aren't using Collectors, you can try the work-in-progress code in the bugfix/44 branch.

I'll keep working on this until I can publish a fully operational release.

**Comment by wwkimball** [comment]
Another update:  I have abandoned branch bugfix/44 because I accidentally put it into a nasty Merge Conflict state.  I moved my work to a new branch, bugfix/44-type_to_class, which is now functionally complete.  I'm just working on code quality issues (mypy, pylint), then documentation at this point.  The Issue #44 is resolved in this branch and I've added new unit tests to demonstrate the fix in several variations.

I'll keep at these code quality issues and release a new version as soon as I can.  Sorry for the long wait.
